README.md
# <img src="https://raw.githubusercontent.com/nephlm/codex-autocommit/main/logo_128.png" width=48> Codex Autocommit

VSCode extension to automatically stage and commit changes on an interval.

## Documentation

This extension is not particularly the best practice for code,
it was built to support using VSCode as a writing platform.

Every ten minutes (configurable) this extension will
add all new and changed files from the git working tree to the index.
Then commit the changeset to the repo, and if applicable,
push to a remote repository.

Which is the detailed version of: *Every ten minutes, it will push all changes to GitHub (or whatever remote git provider you use.)*

### Limitations

* Only works with git.
* Does not support multi-root workspaces.
* It adds, commit and pushes everything in the workspace (unless excluded by .gitignore), do not create files in the workspace with sensitive information (passwords, credit card numbers, etc.) while autocommit is active.

## Getting Started

You still need to initialize your git repo
and configure your remotes as per normal.
It just uses the git configuration in the workspace.

(Github)[https://github.com] has spent a lot of time and resources on explaining version control and how to set it up for your work than I ever will, so instead of trying I'll point to their hard work.

* [What is Git/Version Control](https://git-scm.com/doc)
* [Setting up Git](https://docs.github.com/en/get-started/quickstart/set-up-git)
* [Creating a Repo](https://docs.github.com/en/get-started/quickstart/create-a-repo)
* [Create a Repo from Existing directory](https://docs.github.com/en/get-started/importing-your-projects-to-github/importing-source-code-to-github/adding-locally-hosted-code-to-github)
* [Getting started with Git](https://docs.github.com/en/get-started/getting-started-with-git)
* [Git and VSCode](https://code.visualstudio.com/docs/sourcecontrol/overview)
* [Ignoring Files](https://docs.github.com/en/get-started/getting-started-with-git/ignoring-files)

## Installation

Configure your git in your workspace
by whichever method you normally use.

In VSCode, choose the extension activities
and this extension can be found by searching for `codex-autocommit`.
Click the `install` button, and you're off to the races.

## Configuration

By default, this extension doesn't do anything,
you'll need to at least set `codex-autocommit.active` to `true`
before it will start working.

There are two settings for this extension.

`codex-autocommit.active`  

By default this is off,
and should probably be activated on a per-workspace basis.

`codex-autocommit.intervalMinutes`

How many minutes the extension will sleep between autocommits.

## Feedback

For feature request, comments or bug reports, please create an [issue on Github](https://github.com/nephlm/codex-autocommit/issues)

## License

[MIT](https://choosealicense.com/licenses/mit/)

## Related

Here are some other VSCode extensions I wrote to support my
writing instead of writing.

[Codex Autocommit](https://marketplace.visualstudio.com/items?itemName=ZenBrewismBooks.codex-autocommit&ssr=false#overview) -
Automatically take a snapshot of the manuscript every interval and store it on a remote git server.

[Codex Manuscript File Operations](https://marketplace.visualstudio.com/items?itemName=ZenBrewismBooks.codex-manuscript-file-operations) - A set of file operations to facilitate splitting and merging scenes and to maintaining the fileorder of the manuscript.

[Codex Manuscript Wordcount](https://marketplace.visualstudio.com/items?itemName=ZenBrewismBooks.codex-manuscript-wordcount) - Show whole manuscript word counts and set manuscript and writing session targets.

## Authors

[@nephlm](https://www.github.com/nephlm)

## Logo

by [Lorc](https://game-icons.net/1x1/lorc/letter-bomb.html) under [CC BY 3.0](http://creativecommons.org/licenses/by/3.0/)


CHANGELOG.md
# Change Log

All notable changes to the "codex-autocommit" extension will be documented in this file.

## [0.3.19]

- Add links to git documentation to the README, and otherwise fill in some gaps.

## [0.3.18]

- Added logo to project and README.

## [0.3.0 - 0.3.17]

- Testing releases as I attempted to figure what the marketplace validation found objectionable.  Only real changes were some editing of the readme.

## [0.2.0]

- Initial MVP release


autocommit.js
import * as vscode from 'vscode';

let lastCommitTime: number | undefined;

const commitAndPush = async () => {
    const scm = vscode.scm.activeProvider;
    if (!scm) {
        return;
    }

    const state = scm.state;
    const dirtyFiles = Array.from(state.values())
        .filter(s => s.modified || s.staged || s.conflict);

    if (dirtyFiles.length > 0) {
        const message = `Committing ${dirtyFiles.length} dirty files`;
        const progress = vscode.window.createProgressIndicator();
        progress.title = message;

        progress.begin(dirtyFiles.length);

        // Add dirty files to the staging area
        await scm.inputBox.value = 'Committing dirty files';
        for (const file of dirtyFiles) {
            scm.stage(file.resourceUri);
            progress.increment();
        }

        // Commit the staged changes
        await scm.inputBox.value = 'Committing changes';
        await scm.commit();

        // Push the changes to the remote repository
        await scm.inputBox.value = 'Pushing changes';
        await scm.push();

        progress.done();
        vscode.window.showInformationMessage(`${message} successful`);
    }
};

export function activate(context: vscode.ExtensionContext) {
    // Register a command that can be triggered from the command palette
    const disposable = vscode.commands.registerCommand(
        'extension.commitAndPush',
        commitAndPush
    );
    context.subscriptions.push(disposable);

    // Run the commit and push task every time the SCM state changes
    const scm = vscode.scm.activeProvider;
    if (scm) {
        scm.onDidChangeState(async () => {
            if (!lastCommitTime || Date.now() - lastCommitTime > 1000 * 60 * 10) {
                // It has been more than 10 minutes since the last commit
                lastCommitTime = Date.now();
                await commitAndPush();
            }
        });
    }
}

export function deactivate() {}


extension.js
// The module 'vscode' contains the VS Code extensibility API
// Import the module and reference it with the alias vscode in your code below
const vscode = require("vscode");

// This method is called when your extension is activated
// Your extension is activated the very first time the command is executed

let timeoutId;

const commitAndPush = async () => {
  const gitExtensionHandle = vscode.extensions.getExtension("vscode.git");
  if (!gitExtensionHandle.isActive) {
    console.log("git extension not active, activating");
    gitExtensionHandle.activate().then(() => {
      console.log("git extension activated");
      timeoutId = setTimeout(commitAndPush, 10 * 1000);
    });
    console.log("Sleeping for 10 seconds.");
    return;
  }

  const workspaceConfig = vscode.workspace.getConfiguration("codex-autocommit");
  console.log("Codex Autocommit active: " + workspaceConfig.get("active"));
  console.log(
    "Codex Autocommit interval: " +
      workspaceConfig.get("intervalMinutes") +
      "min"
  );
  if (!workspaceConfig.get("active")) {
    console.log("Codex Autocommit is not active; Exiting");
    return;
  }

  const api = gitExtensionHandle.exports.getAPI(1);
  console.log("git extension state is: " + api.state);
  if (api.state != "initialized") {
    console.log("git extension not initialized");
    timeoutId = setTimeout(commitAndPush, 10 * 1000);
    console.log("Sleeping for 10 seconds");
    return;
  }

  const setNext = function () {
    const sleepMinutes = workspaceConfig.get("intervalMinutes");
    timeoutId = setTimeout(commitAndPush, sleepMinutes * 60 * 1000);
    console.log(
      "commitAndPush complete, sleeping for " + sleepMinutes + " minutes"
    );
  };

  const check_push = function (repository) {
    if (repository.state.remotes.length) {
      console.log("Pushing Changes");
      var push_promise = repository.push();
      push_promise.then(() => {
        setNext();
      });
    } else {
      console.log("No remote to push to.");
      setNext();
    }
  };

  const check_commit = function (repository) {
    if (repository.state.indexChanges.length) {
      console.log("Committing changes.");
      const date = new Date();
      const iso_string = date.toISOString();
      var commitPromise = repository.commit(
        iso_string + ": " + repository.inputBox.value
      );
      commitPromise.then(check_push(repository));
    } else {
      console.log("Nothing to commit.");
      check_push(repository);
    }
  };

  const repository = api.repositories[0];
  await repository.status();

  const modified = repository.state.workingTreeChanges.map((item) => {
    return item.uri.path;
  });
  var addPromise;
  console.log(modified);

  if (modified.length) {
    console.log("Staging Files");
    addPromise = repository.add(modified);
    addPromise.then(() => check_commit(repository));
  } else {
    console.log("Working Tree is clean.");
    check_commit(repository);
  }
};

/**
 * @param {vscode.ExtensionContext} context
 */
function activate(context) {
  // Use the console to output diagnostic information (console.log) and errors (console.error)
  // This line of code will only be executed once when your extension is activated
  console.log(
    'Congratulations, your extension "codex-autocommit" is now active!'
  );

  // Run the commit and push task every time the SCM state changes
  // const scm = vscode.scm.activeSourceControl;

  const commitDisposable = vscode.commands.registerCommand(
    "codex-autocommit.commitAndPush",
    () => {
      commitAndPush();
      console.log("commit and push manual command called");
    }
  );
  context.subscriptions.push(commitDisposable);

  context.subscriptions.push(
    vscode.workspace.onDidChangeConfiguration((e) => {
      if (
        e.affectsConfiguration("codex-autocommit.active") ||
        e.affectsConfiguration("codex-autocommit.intervalMinutes")
      ) {
        console.log("Configuraton changed; restarting");
        if (timeoutId) {
          clearTimeout(timeoutId);
        }
        timeoutId = setTimeout(commitAndPush, 10 * 1000);
      }
    })
  );

  commitAndPush();
}

// This method is called when your extension is deactivated
function deactivate() {}

module.exports = {
  activate,
  deactivate,
};
